### ISSUE TITLE
update source

### ISSUE THREAD

**Issue by BlackScript** [flo199213/APEROLiker#1]
Dependencies / Libaries for current adurino studio (v higher then 2)  no successful build possible
webserver not work when use the fix for arduino higher v3 which deployed on https://github.com/me-no-dev/ESPAsyncWebServer
change from spiffs to littlefs possible ? so upload is over the littlefs uploader plugin possible

**Comment by flo199213** [comment]
Hi BlackScript,

thanks for your interest in my APEROLiker firmware!

I am running into the same problem to build the firmware with the latest dependencies.

I use the latest Arduino IDE:
Version: 2.3.2
Date: 2024-02-20T10:04:35.814Z

I use the latest version of arduino-esp32:
Using board 'esp32s2': 3.1.0-RC1  (Contains esp-idf v5.3)
Using core 'esp32':    3.1.0-RC1  (Contains esp-idf v5.3)

I use the latest version of each particular library:
  -> [SPI@3.1.0]
  -> [SPIFFS@3.1.0]
  -> [FS@3.1.0]
  -> [WiFi@3.1.0]
  -> [Networking@3.1.0]
  -> [Adafruit ST7735 and ST7789 Library@1.10.4]
  -> [Adafruit GFX Library@1.11.10]
  -> [Adafruit BusIO@1.16.1] (Adafruit_I2CDevice.h)
  -> [Wire@3.1.0]
  -> [Preferences@3.1.0]
  -> [ESPmDNS@3.1.0]
  -> [AsyncTCP@1.1.4]
  -> [ESPAsyncWebServer@3.1.0]  <------- Compile error on "undefined reference to `pxCurrentTCB'"

As my research has shown, this is because the new esp-idf v5.3 has updated the dependencies for FreeRTOS. Unfortunately, as always, this is typical for Espressif - no downward compatibility.

In the short term I sadly don't see any possibility of switching from SPIFFS to LittleFS for personal time reasons, as I'm not yet familiar with LittleFS. 


Uncommenting the following lines in "ESPAsyncWebServer/src/AsyncWebSynchronization.h" will do the trick:

```
bool lock() const {
  /*extern void *pxCurrentTCB;
  if (_lockedBy != pxCurrentTCB) {
    xSemaphoreTake(_lock, portMAX_DELAY);
    _lockedBy = pxCurrentTCB;
    return true;
  }*/
  return false;
}
```
https://github.com/me-no-dev/ESPAsyncWebServer/blob/7f3753454b1f176c4b6d6bcd1587a135d95ca63c/src/AsyncWebSynchronization.h#L28

So now what's the best thing to do?
A bug fix should actually be sought for the ESPAsyncWebServer... But I don't know enough about FreeRTOS to say how to correctly replace the variable "pxCurrentTCB" in esp-idf v5.3.

**Comment by flo199213** [comment]
Hi BlackScript,

thanks for your interest in my APEROLiker firmware!

I am running into the same problem to build the firmware with the latest dependencies.

I use the latest Arduino IDE:
Version: 2.3.2
Date: 2024-02-20T10:04:35.814Z

I use the latest version of arduino-esp32:
Using board 'esp32s2': 3.1.0-RC1  (Contains esp-idf v5.3)
Using core 'esp32':    3.1.0-RC1  (Contains esp-idf v5.3)

I use the latest version of each particular library:
  -> [SPI@3.1.0]
  -> [SPIFFS@3.1.0]
  -> [FS@3.1.0]
  -> [WiFi@3.1.0]
  -> [Networking@3.1.0]
  -> [Adafruit ST7735 and ST7789 Library@1.10.4]
  -> [Adafruit GFX Library@1.11.10]
  -> [Adafruit BusIO@1.16.1] (Adafruit_I2CDevice.h)
  -> [Wire@3.1.0]
  -> [Preferences@3.1.0]
  -> [ESPmDNS@3.1.0]
  -> [AsyncTCP@1.1.4]
  -> [ESPAsyncWebServer@3.1.0]  <------- Compile error on "undefined reference to `pxCurrentTCB'"

As my research has shown, this is because the new esp-idf v5.3 has updated the dependencies for FreeRTOS. Unfortunately, as always, this is typical for Espressif - no downward compatibility.

In the short term I sadly don't see any possibility of switching from SPIFFS to LittleFS for personal time reasons, as I'm not yet familiar with LittleFS. 


Uncommenting the following lines in "ESPAsyncWebServer/src/AsyncWebSynchronization.h" will do the trick:

```
bool lock() const {
  /*extern void *pxCurrentTCB;
  if (_lockedBy != pxCurrentTCB) {
    xSemaphoreTake(_lock, portMAX_DELAY);
    _lockedBy = pxCurrentTCB;
    return true;
  }*/
  return false;
}
```
https://github.com/me-no-dev/ESPAsyncWebServer/blob/7f3753454b1f176c4b6d6bcd1587a135d95ca63c/src/AsyncWebSynchronization.h#L28

So now what's the best thing to do?
A bug fix should actually be sought for the ESPAsyncWebServer... But I don't know enough about FreeRTOS to say how to correctly replace the variable "pxCurrentTCB" in esp-idf v5.3.

**Comment by BlackScript** [comment]
Hey. i currently got a nearly working build when using this fork of ESPAsyncWebServer https://github.com/mathieucarbou/ESPAsyncWebServer and https://github.com/mathieucarbou/AsyncTCP/releases/tag/v3.2.0
only thing is that there is a problem with webhook so it reboots when changing values over web ... but there seems already some possible fixes available in the fork issues. so this seems currently the best way with the newest ide and so on...

**Comment by flo199213** [comment]
Ah nice. The fork from mathieucarbou actually seems to be much better maintained.

To be honest, the WiFi functionality of the APEROLiker is only a proof of concept and was intended as a gimmick (nice to have). This is why the build switch ‘#define WIFI_MIXER’ exists. Without WiFi the mixer is maximally stable.
https://github.com/flo199213/APEROLiker/blob/5fef622786bab3b6e0d7de520786b7b43c873128/ESP32S2_Aperoliker_V1.2/Config.h#L36

Of course, that doesn't mean that I wouldn't be interested in getting the firmware with WiFi just as stable... 👍

**Comment by BlackScript** [comment]
yes. already seen this ... but i like the gimmick :D which sure not mean this kills this cool device for me ... :) 
[esp32.esp32.esp32s2.zip](https://github.com/user-attachments/files/16840791/esp32.esp32.esp32s2.zip)
here is the build with the current arduino studio and newest board esp32 v 3.0.4 if you want to try this version

**Comment by flo199213** [comment]
There is now also a new version V1.3, which no longer relies on any asynchronous TCP libraries. I was tired of constant incompatibilities and that the solution did not build. Please have a look if this one is suitable for you: [CocktailCube](https://github.com/flo199213/CocktailCube). The firmware also has many other improvements, such as being able to select the design/recipe via the settings and everything is easier to customize.
